File: C++/OrderProcessor.cpp
// OrderProcessor — SOLID violations demo
#include <iostream>
#include <string>
#include <vector>
using namespace std;

// =======================
// BAD: Violates SRP, OCP, DIP
// =======================
class OrderProcessor {
public:
    void processOrder(const string& customerType, const vector<string>& items, const string& paymentMethod) {

        // 1. Calculate total
        double total = 0;
        for (const auto& item : items) {
            if (item == "LAPTOP")      total += 80000;
            else if (item == "PHONE")  total += 30000;
            else if (item == "HEADPHONES") total += 5000;
        }

        // 2. Apply discount (OCP violation)
        if (customerType == "PREMIUM")     total *= 0.80;
        else if (customerType == "REGULAR") total *= 0.95;

        // 3. Process payment (OCP violation)
        if (paymentMethod == "CREDIT_CARD")
            cout << "Processing credit card payment of Rs." << total << "\n";
        else if (paymentMethod == "UPI")
            cout << "Processing UPI payment of Rs." << total << "\n";
        else if (paymentMethod == "NETBANKING")
            cout << "Processing NetBanking payment of Rs." << total << "\n";

        // 4. Save to DB (SRP violation)
        cout << "Saving order to database...\n";

        // 5. Send email (SRP violation)
        cout << "Sending confirmation email to customer...\n";

        // 6. Generate invoice (SRP violation)
        cout << "=== INVOICE ===\n";
        cout << "Total after discount: Rs." << total << "\n";
        cout << "Payment via: " << paymentMethod << "\n";
    }
};


// =======================
// BAD: LSP Violation
// =======================
class Animal {
public:
    virtual string sound() { return "Some sound"; }
    virtual string fly()   { return "I can fly!"; }
    virtual ~Animal() {}
};

class Dog : public Animal {
public:
    string fly() override { throw runtime_error("Dogs can't fly!"); }
};

class Bird : public Animal {
public:
    string fly() override { return "I am flying!"; }
};


// =======================
// BAD: ISP Violation
// =======================
class WorkerInterface {
public:
    virtual void work()          = 0;
    virtual void eat()           = 0;
    virtual void sleep()         = 0;
    virtual void attendMeeting() = 0;
    virtual void writeCode()     = 0;
    virtual ~WorkerInterface() {}
};

class HumanWorker : public WorkerInterface {
public:
    void work()          override { cout << "Human working\n"; }
    void eat()           override { cout << "Human eating\n"; }
    void sleep()         override { cout << "Human sleeping\n"; }
    void attendMeeting() override { cout << "Human in meeting\n"; }
    void writeCode()     override { cout << "Human writing code\n"; }
};

class RobotWorker : public WorkerInterface {
public:
    void work()          override { cout << "Robot working\n"; }
    void eat()           override { throw runtime_error("Robots don't eat!"); }
    void sleep()         override { throw runtime_error("Robots don't sleep!"); }
    void attendMeeting() override { cout << "Robot in meeting\n"; }
    void writeCode()     override { cout << "Robot writing code\n"; }
};


// =======================
// GOOD: DIP — depend on abstraction
// =======================
class IDatabase {
public:
    virtual void save(const string& data) = 0;
    virtual ~IDatabase() {}
};

class MySQLDatabase : public IDatabase {
public:
    void save(const string& data) override {
        cout << "Saving to MySQL: " << data << "\n";
    }
};

class NoSQLDatabase : public IDatabase {
public:
    void save(const string& data) override {
        cout << "Saving to NoSQL: " << data << "\n";
    }
};

class ReportService {
    IDatabase* database;
public:
    ReportService(IDatabase* db) : database(db) {}

    void setDatabase(IDatabase* db) { database = db; }

    void generateReport(const string& data) {
        cout << "Generating report...\n";
        database->save(data);
    }
};


int main() {
    OrderProcessor processor;
    processor.processOrder("PREMIUM", {"LAPTOP", "PHONE"}, "UPI");

    MySQLDatabase mysql;
    ReportService report(&mysql);
    report.generateReport("Order #1001");

    NoSQLDatabase nosql;
    report.setDatabase(&nosql);
    report.generateReport("Order #1002");

    return 0;
}

File: Python/OrderProcessor.py
# OrderProcessor — SOLID violations demo
# Shows bad patterns and refactored good patterns

from abc import ABC, abstractmethod


# =======================
# BAD: OrderProcessor violates SRP, OCP, DIP
# =======================
class OrderProcessor:
    def process_order(self, customer_type: str, items: list, payment_method: str):

        # 1. Calculate total
        total = 0
        for item in items:
            if item == "LAPTOP":      total += 80000
            elif item == "PHONE":     total += 30000
            elif item == "HEADPHONES": total += 5000

        # 2. Apply discount — hardcoded (OCP violation)
        if customer_type == "PREMIUM":
            total *= 0.80
        elif customer_type == "REGULAR":
            total *= 0.95

        # 3. Process payment — hardcoded (OCP violation)
        if payment_method == "CREDIT_CARD":
            print(f"Processing credit card payment of ₹{total}")
        elif payment_method == "UPI":
            print(f"Processing UPI payment of ₹{total}")
        elif payment_method == "NETBANKING":
            print(f"Processing NetBanking payment of ₹{total}")

        # 4. Save to DB (SRP violation)
        print("Saving order to database...")

        # 5. Send email (SRP violation)
        print("Sending confirmation email to customer...")

        # 6. Generate invoice (SRP violation)
        print("=== INVOICE ===")
        print(f"Items: {items}")
        print(f"Total after discount: ₹{total}")
        print(f"Payment via: {payment_method}")


# =======================
# BAD: LSP Violation
# =======================
class Animal:
    def sound(self) -> str:
        return "Some sound"

    def fly(self) -> str:
        return "I can fly!"

class Dog(Animal):
    def fly(self) -> str:
        raise NotImplementedError("Dogs can't fly!")  # LSP violation

class Bird(Animal):
    def fly(self) -> str:
        return "I am flying!"


# =======================
# BAD: ISP Violation
# =======================
class WorkerInterface(ABC):
    @abstractmethod
    def work(self): pass
    @abstractmethod
    def eat(self): pass
    @abstractmethod
    def sleep(self): pass
    @abstractmethod
    def attend_meeting(self): pass
    @abstractmethod
    def write_code(self): pass

class HumanWorker(WorkerInterface):
    def work(self): print("Human working")
    def eat(self): print("Human eating")
    def sleep(self): print("Human sleeping")
    def attend_meeting(self): print("Human in meeting")
    def write_code(self): print("Human writing code")

class RobotWorker(WorkerInterface):
    def work(self): print("Robot working")
    def eat(self): raise NotImplementedError("Robots don't eat!")   # ISP violation
    def sleep(self): raise NotImplementedError("Robots don't sleep!") # ISP violation
    def attend_meeting(self): print("Robot in meeting")
    def write_code(self): print("Robot writing code")


# =======================
# GOOD: DIP — depend on abstraction
# =======================
class IDatabase(ABC):
    @abstractmethod
    def save(self, data: str): pass

class MySQLDatabase(IDatabase):
    def save(self, data: str):
        print(f"Saving to MySQL: {data}")

class NoSQLDatabase(IDatabase):
    def save(self, data: str):
        print(f"Saving to NoSQL: {data}")

class ReportService:
    def __init__(self, database: IDatabase):
        self.database = database

    def set_database(self, database: IDatabase):
        self.database = database

    def generate_report(self, data: str):
        print("Generating report...")
        self.database.save(data)


if __name__ == "__main__":
    processor = OrderProcessor()
    processor.process_order("PREMIUM", ["LAPTOP", "PHONE"], "UPI")

    report = ReportService(MySQLDatabase())
    report.generate_report("Order #1001")

    report.set_database(NoSQLDatabase())
    report.generate_report("Order #1002")
